### DISCUSSION TITLE
How to create a "Flexible template"?

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#1297] [Help] [answered]
TL;DR: I need a system in which changing the template changes the copies, but the copies also have their own specific information.
In other words, it's like a cloned branch, but with only one-way synchronization (instead of bi-directional syncing).
---

Please, I need suggestions for this use case:
My concrete situation is that I will need to deal with a bunch of documents that follow a general template, but also have their own content. I will build the template as I build the documents, so I need that the documents auto update as I change the template. 

In FP, it would be a setup like this:

I have a branch that is a template. From it, I create other branches (I will call those, "copies"):
![image](https://github.com/freeplane/freeplane/assets/77707706/fd887d9c-03f9-4716-ac84-b3ba62481af2)

If I change something in the template, I want it to also change in the copies. For example, adding "blabla 2" in Part 1 in the template results in it being added to all copies:
![image](https://github.com/freeplane/freeplane/assets/77707706/295cb90b-545c-4f90-87a3-946f3e29065f)

But, I also want the choice to have nodes that are specific for each Document, and don't appear on other Documents (and I prefer that they don't appear on the template, but it's not essential).  So, if I add "blabla 3" to Document 1, it ends up like this:
![image](https://github.com/freeplane/freeplane/assets/77707706/c6a5568e-8538-4b3a-bc6e-9a964b3237c1)

It would also be interesting, although not essential, to choose some of the nodes of the template to not appear in the copy (in other words, a opt-out). For example, choosing not to show Part 2 in Document 1, would result in this:
![image](https://github.com/freeplane/freeplane/assets/77707706/8f464d15-7163-4959-9b01-83eda2f31373)

In FP, it's not straightforward, because cloning a branch forces a "full synch" between the cloned branches.

 So, I tried a quick solution using spreedsheets (the first column is the template, and the other columns are the documents; filter the first column and the desired "document-collumn", and you get the final text). To my surprise, neither Notion or Google sheets can take text and images in the cell at the same time (I'm astonished). I need text and images in the cell at the same time.
 A table in LibreOffice Writer can take image and text in the cells, but there is no filtering and view controls, which makes it not usable, because the amount of content is too big.

Now, back to FP, a have a vague idea about using attributes in order to choose which nodes appear on each clone branch, by making the attribute activate a conditional style.

Is there an easier or more obvious solution that I'm not seeing? In FP or any other software.

Thanks,

map used in the example: [flexible template.zip](https://github.com/freeplane/freeplane/files/11994066/flexible.template.zip)

**Comment by euu2021** [comment]
Maybe, this is an easier solution: placing everything inside the "template", and then using attributes to filter only desired nodes.

So, to solve the example above, I would have only the template node, and then:

-If I want to add "blabla 3" only to Document 1, I put on node "blabla 3" the attribute [just show up on | Document 1]

-If want Part 2 to not show up on Document 1, I put on node "Document 1" the attribute [don't show up on | Document 1]

So, when I want to focus on Document 1, I use the filter:
-exclude nodes where "just show up on" exists and is not "Document 1"
-exclude nodes with the attribute [don't show up on | Document 1]
-exclude nodes that have any ancestor with attribute 

![image](https://github.com/freeplane/freeplane/assets/77707706/552c547b-3a0a-489b-bb3e-94ab5d487956)

**Comment by macmarrum** [comment, marked as answer]
This is an interesting topic. I would use single-node clones and a script to keep them added after new ones appear in the template.

Please try [syncFollowersWithTemplates.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/syncFollowersWithTemplates.groovy)

In each follower (your `Document #` nodes) add an attribute `template` with node id of the template.
You can add more then one such attribute to follow many templates at the same time.

You need to run the script each time you add new nodes to the template to have them appear in the followers.

**Reply by euu2021** [reply]
Brilliant! Thanks!
I like this solution more than using attributes and filters, because it allows leaving the branches separated, so I don't need to mess with filters. And it allows some interesting things that I did not think before:

### -template inheritance:
So, if the user wants to create food recipes, he can have:
- an upper template for recipes in general; 
- a more specific template for cakes recipes. This template follows the general food recipe.
- and then the actual chocolate cake can simply follow the cake template, because it already brings the general recipe template.

### -multi template 
Continuing the chocolate cake example, it can also profit from the template "chocolate mousse recipe". So, it follows both "cake recipe", and "chocolate mousse recipe"


This will be useful to me right now, because my documents are grouped in categories, and each category can have it's own template.



---
### -opt out 
Now, to achieve the opt out, one way is :
- add a node with the text "out" as a child of the node that is to be excluded
- create a conditional style for the condition "child has text 'out'". The applied style makes it clear that this node is of no interest for that context. Luckily, the conditional style is not applied to the other clones:
![image](https://github.com/freeplane/freeplane/assets/77707706/1959fae7-4a3f-4eb9-8ed3-5482111586a7)



### -A shortcoming of the script: 
When nodes are deleted on the template, it's kept on the follower. 
Automatically deleting it on the follower would be undesirable, because it may contain relevant information that the user may not have considered when deleting the node on the template. Maybe, a solution would be:
- the nodes cloned in the follower, from the template, have the attribute "status"
- if the node is currently in the template, too, then the status is "active"
- if the node is not in the template (ie, it was deleted on the template), then the status is "deprecated"

This would allow the user to create a conditional style that makes clear that a node was deleted on the template. Ie, a conditional style that gets triggered by the attribute [status|deprecated].

**Reply by macmarrum** [reply]
> Brilliant! Thanks!
I like this solution more than using attributes and filters, because it allows leaving the branches separated, so I don't need to mess with filters. And it allows some interesting things that I did not think before

:heart: 

> -opt out

Another approach is to use a template (cloned from the original) without the opt-out node on it.

> -A shortcoming of the script:
> When nodes are deleted on the template, it's kept on the follower.

To delete all clones of a node (in the template and in all followers)
1. Select the node
2. In the Filter toolbar, choose `Clones of selection`, then `Filter->Select all matching nodes`
3. Delete

If this is something that you do often, it can be made into a script.

**Reply by euu2021** [reply]
I came up with a quick fix to deprecate the deleted nodes. This script, to be used as the Delete key, works like that:
- checks if the node is inside the template branch (I will keep all my templates inside a branch with ID_866298839)
- if yes, then change the node style to DEPRECATEDNode and then delete it

```
// @ExecutionModes({ON_SELECTED_NODE})

def nodeList = [ ]
nodeList = node.getPathToRoot().collect()

def selectedNodes = [ ]
selectedNodes = c.selecteds.collect()

if (nodeList.contains(ID_866298839)) {
    selectedNodes.each{it.style.setName("DEPRECATEDNode")
    }
    menuUtils.executeMenuItems([ 'DeleteAction' ])
}

else {
    menuUtils.executeMenuItems([ 'DeleteAction' ])
}
```

**Reply by macmarrum** [reply]
> When nodes are deleted on the template, it's kept on the follower. Automatically deleting it on the follower would be undesirable, because it may contain relevant information that the user may not have considered when deleting the node on the template. Maybe, a solution would be:
> 
>     * the nodes cloned in the follower, from the template, have the attribute "status"
> 
>     * if the node is currently in the template, too, then the status is "active"
> 
>     * if the node is not in the template (ie, it was deleted on the template), then the status is "deprecated"
> 
> 
> This would allow the user to create a conditional style that makes clear that a node was deleted on the template. Ie, a conditional style that gets triggered by the attribute [status|deprecated].

You can tell that a node is not following a template counterpart by the fact that it isn't a clone. When `Preferences…->Appearance->Clones->Paint clone marks` is enabled, the lack of the clone indicator (a small square at the edge-node joint) informs about it.

**Reply by euu2021** [reply]
> I came up with a quick fix to deprecate the deleted nodes. This script, to be used as the Delete key, works like that:

This solution is still very limited though, because there is also a common situation where I rearrange the nodes in the template (the strong suit of FP is rearranging nodes), and the result is this (in the template, I moved Part 1 to be inside Chapter 1):

![image](https://github.com/freeplane/freeplane/assets/77707706/9b1c6389-5d14-41d3-a718-3f15370577f7)

**Reply by macmarrum** [reply]
I think this situation could be handled; I'm going to improve the script – something like: search for a node inside the follower branch and move it to the correct parent instead of expecting it to be at the same place as the template.

**Comment by euu2021** [comment]
> You can tell that a node is not following a template counterpart by the fact that it isn't a clone. When `Preferences…->Appearance->Clones->Paint clone marks` is enabled, the lack of the clone indicator (a small square at the edge-node joint) informs about it.

But, when there are multiple followers of the template, even if the node is deleted in the template, multiple clones remain, so the clone indicator will remain, too:
![image](https://github.com/freeplane/freeplane/assets/77707706/8ef490d0-5a20-4cc4-8695-fde1ad971d83)

**Reply by macmarrum** [reply]
True.

>  came up with a quick fix to deprecate the deleted nodes. This script, to be used as the Delete key, works like that:
>
>  - checks if the node is inside the template branch (I will keep all my templates inside a branch with ID_866298839)
>  - if yes, then change the node style to DEPRECATEDNode and then delete it

This is a great idea :+1:

**Comment by macmarrum** [comment]
> > I came up with a quick fix to deprecate the deleted nodes. This script, to be used as the Delete key, works like that:
> 
> This solution is still very limited though, because there is also a common situation where I rearrange the nodes in the template (the strong suit of FP is rearranging nodes), and the result is this (in the template, I moved Part 1 to be inside Chapter 1):
> 
> ![image](https://user-images.githubusercontent.com/77707706/252162069-9b1c6389-5d14-41d3-a718-3f15370577f7.png)

----

> I think this situation could be handled; I'm going to improve the script – something like: search for a node inside the follower branch and move it to the correct parent instead of expecting it to be at the same place as the template.

---

The script has been updated: [syncFollowersWithTemplates.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/syncFollowersWithTemplates.groovy)

It support the moving of nodes in the template branch. Also, template attribute can now be a hyperlink to the template node.

`Ctrl+F5` might be required to force a refresh of the page.

**Reply by euu2021** [reply]
I did some tests, and the script seems to do everything that I need. I will introduce it in my workflow. Thanks!

**Reply by euu2021** [reply]
A small addition that I found useful: "tagging" the nodes from the template with the attribute [isFromTemplate?|yes].

This allowed me to do two things:
- trigger a conditional style for those nodes inside the follower, in order to make it visually clear that those nodes are from template
- block deleting those nodes in the follower branch. As the script forces full synch from the template, deleting the node on the follower is useless, and the user can get confused when nodes reappear even after getting deleted. I may even try to create a listener to block the user from even moving those nodes. The idea is to make those template nodes get rock solid in place.

So, the tagging is done by adding this to your script (again, all my templates are inside node ID_866298839):
```
mindMap.root.findAll().each {
    if (it.isDescendantOf(ID_866298839)) {
        it["isFromTemplate?"] = "yes"
    }
    else {
    }
}
```

And to block the deletion of those nodes, I changed [my script](https://github.com/freeplane/freeplane/discussions/1297#discussioncomment-6397509) above to be like this:

```
import javax.swing.JOptionPane

// @ExecutionModes({ON_SELECTED_NODE})

def nodeList = [ ]
nodeList = node.getPathToRoot().collect()

def selectedNodes = [ ]
selectedNodes = c.selecteds.collect()

if (nodeList.contains(ID_866298839)) {
    selectedNodes.each{it.style.setName("DEPRECATEDNode")
    }
    menuUtils.executeMenuItems([ 'DeleteAction' ])
}

else {
    selectedNodes.each{
        if (it.attributes.findFirst("isFromTemplate?") == -1) {
            it.delete()
        } else {
            JOptionPane.showMessageDialog(null, "This node is from template")
        }
    }
}
```

**Reply by macmarrum** [reply]
I renamed the script to a more generic [syncFollowerBranchesWithMasterBranches.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/syncFollowerBranchesWithMasterBranches.groovy)

> A small addition that I found useful: "tagging" the nodes from the template with the attribute [isFromTemplate?|yes].

A good idea :bulb::+1: 

I have extended the script to optionally tag nodes with `follower/master` (switched off by default). It can be renamed and enabled in the script.
Why not `isFromTemplate?`? Because the nodes are clones and therefore the same attribute is on the master as on the followers.

> block deleting those nodes in the follower branch. As the script forces full sync from the template, deleting the node on the follower is useless, and the user can get confused when nodes reappear even after getting deleted. 
> [...]
> And to block the deletion of those nodes, I changed https://github.com/freeplane/freeplane/discussions/1297#discussioncomment-6397509 above to be like this:

I think there is some room for improvement as it comes to the script, but the idea to identify a follower node and disallow deletion is great :rocket:
Let me know if you get stuck – I'll be glad to lend a hand :smile::hand:

**Reply by macmarrum** [reply]
> I renamed the script to a more generic [syncFollowerBranchesWithMasterBranches.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/syncFollowerBranchesWithMasterBranches.groovy)

I also and changed the attribute name in the script from `template` to `master.branch`.

**Reply by euu2021** [reply]
> I have extended the script to optionally tag nodes with follower/master (switched off by default). It can be renamed and enabled in the script.

Nice. It got easier than relying on the node Id as I did above.

I won't be able to try to play with node change listeners in the next couple of weeks, because my time will be short. But, I will keep it on my wishlist. Maybe, the most interesting behavior is that any changes in the master or in the follower triggers the script, so there is no need to reapply the script manually And it would also, as a result, block changes in "template nodes" inside the follower. I'm not sure if this automatic behavior will get wild and cause undesired effects, though 🤔
